src/huggingface_hub/file_download.py

```diff
@@ -1130,19 +1130,21 @@ def _hf_hub_download_to_cache_dir(
     # In that case store a ref.
     _cache_commit_hash_for_specific_revision(storage_folder, revision, commit_hash)
 
-    # If file already exists, return it (except if force_download=True)
-    if not force_download:
-        if os.path.exists(pointer_path):
-            return pointer_path
-
-        if os.path.exists(blob_path):
-            # we have the blob already, but not the pointer
-            _create_symlink(blob_path, pointer_path, new_blob=False)
-            return pointer_path
-
     # Prevent parallel downloads of the same file with a lock.
     # etag could be duplicated across repos,
     lock_path = os.path.join(locks_dir, repo_folder_name(repo_id=repo_id, repo_type=repo_type), f"{etag}.lock")
+    Path(lock_path).parent.mkdir(parents=True, exist_ok=True)
+
+    # If file already exists, return it (except if force_download=True)
+    with WeakFileLock(lock_path):
+        if not force_download:
+            if os.path.exists(pointer_path):
+                return pointer_path
+
+            if os.path.exists(blob_path):
+                # we have the blob already, but not the pointer
+                _create_symlink(blob_path, pointer_path, new_blob=False)
+                return pointer_path
```

Comment:
Two suggestions here:
- The pointer file is immutable once created, if it already exists, we can return safely without grabing the lock. 
- I think it's safer to double-check again that the `pointer_path` does not exist (line 1146 in my code suggestion) because another process can create the pointer in the interval between the outer if (` if not force_download and os.path.exists(blob_path)`) and the moment we acquired the lock, it's rare but stiil let's cover this race window.

```suggestion
 
    # pointer already exists -> immediate return
    if not force_download and os.path.exists(pointer_path):
        return pointer_path

    # Blob exists but pointer must be (safely) created -> take the lock
    if not force_download and os.path.exists(blob_path):
        with WeakFileLock(lock_path):
            if not os.path.exists(pointer_path):
                _create_symlink(blob_path, pointer_path, new_blob=False)
            return pointer_path
```

Reply:
(this suggestion should fix the failing test)